### DISCUSSION TITLE
Testing against the Moddable Two board

### DISCUSSION THREAD

**Discussion by rei-vilo** [phoddie/node-red-mcu#46] [Q&A]
I ordered the [Moddable Two](https://www.moddable.com/moddable-two.php) board and rushed to test it when I received it. Very nice board with a crisp display.

I followed the procedure at [Moddable Two Developer Guide](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/devices/moddable-two.md) to install the [Moddable SDK](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/Moddable%20SDK%20-%20Getting%20Started.md) and the [ESP32 tools](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/devices/esp32.md). 

## Tests with `mcconfig` succeed

I am using the minimal example.

```
cd $MODDABLE/examples/piu/balls
mcconfig -d -m -p esp32/moddable_two
```

The example builds and links, uploads and runs fine.

## Tests with `node-red-mcu-plugin` fail

I am using the minimal example.

<img width="563" alt="Modabble" src="https://user-images.githubusercontent.com/951776/204099056-332bf192-b7f7-4306-a1e9-cbcedac7e3e6.png">

<details>
  <summary>Code</summary>
  
```
[
    {
        "id": "9255b5caec02d914",
        "type": "inject",
        "z": "104f181135f2ef9c",
        "name": "",
        "props": [
            {
                "p": "payload"
            },
            {
                "p": "topic",
                "vt": "str"
            }
        ],
        "repeat": "",
        "crontab": "",
        "once": false,
        "onceDelay": 0.1,
        "topic": "",
        "payload": "",
        "payloadType": "date",
        "_mcu": {
            "mcu": true
        },
        "x": 180,
        "y": 80,
        "wires": [
            [
                "6c7b846e3025a90b"
            ]
        ]
    },
    {
        "id": "6c7b846e3025a90b",
        "type": "debug",
        "z": "104f181135f2ef9c",
        "name": "Debug from MCU",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "payload",
        "targetType": "msg",
        "statusVal": "",
        "statusType": "auto",
        "_mcu": {
            "mcu": true
        },
        "x": 510,
        "y": 200,
        "wires": []
    }
]
```

</details>

Selected board is **esp32/moddable-two**. Build fails with 

> /Users/USER/esp32/esp-idf/export.sh: line 99: python: command not found

<details>
  <summary>Log</summary>

```
Starting Build process...
Creating build environment for platform esp32/moddable_two.
Working directory: /Users/USER/.node-red/mcu-plugin-cache/configtgrufvjolkj
$IDF_PATH identified: /Users/USER/esp32/esp-idf
> cd /Users/USER/.node-red/mcu-plugin-cache/configtgrufvjolkj
> source $IDF_PATH/export.sh && mcconfig -d -x localhost:5004 -m -p esp32/moddable_two -t build ssid="LAN-0A61FC7B" password="MikNycmJdqAeNsmHsryA6x3t"
Detecting the Python interpreter
Checking "python" ...
/Users/USER/esp32/esp-idf/toUSER/detect_python.sh: line 16: python: command not found
Checking "python3" ...
Python 3.9.6
"python3" has been detected
Adding ESP-IDF toUSER to PATH...
Using Python interpreter in 
Checking if Python packages are up to date...
/Users/USER/esp32/esp-idf/export.sh: line 99: python: command not found
```

</details>

Python 3 has been detected but line 99 is hard-coded and calls `python` instead.

Changing line /Users/USER/esp32/esp-idf/export.sh: line 99 from `python` to `python3`

``` 
python3 "${IDF_PATH}/tools/check_python_dependencies.py" || return 1
```

Build goes further with fails anew.

> IDF_PYTHON_ENV_PATH: (not set)

<details>
  <summary>Log</summary>

```
Starting Build process...
Creating build environment for platform esp32/moddable_two.
Working directory: /Users/USER/.node-red/mcu-plugin-cache/configtgrufvjolkj
$IDF_PATH identified: /Users/USER/esp32/esp-idf
> cd /Users/USER/.node-red/mcu-plugin-cache/configtgrufvjolkj
> source $IDF_PATH/export.sh && mcconfig -d -x localhost:5004 -m -p esp32/moddable_two -t build ssid="LAN-0A61FC7B" password="MikNycmJdqAeNsmHsryA6x3t"
Detecting the Python interpreter
Checking "python" ...
/Users/USER/esp32/esp-idf/toUSER/detect_python.sh: line 16: python: command not found
Checking "python3" ...
Python 3.9.6
"python3" has been detected
Adding ESP-IDF toUSER to PATH...
Using Python interpreter in 
Checking if Python packages are up to date...
The following Python requirements are not satisfied:
click>=7.0
pyserial>=3.3
cryptography>=2.1.4
pyparsing>=2.0.3,<2.4.0
pyelftoUSER>=0.22
idf-component-manager~=1.0
gdbgui==0.13.2.0
pygdbmi<=0.9.0.2
python-socketio<5
jinja2<3.1  # See https://github.com/espressif/esp-idf/issues/8760
itsdangerous<2.1
kconfiglib==13.7.1
reedsolo>=1.5.3,<=1.5.4
bitstring>=3.1.6
ecdsa>=0.16.0
construct==2.10.54
Please follow the instructions found in the "Set up the toUSER" section of ESP-IDF Getting Started Guide
Diagnostic information:
    IDF_PYTHON_ENV_PATH: (not set)
    Python interpreter used: /Applications/Xcode.app/Contents/Developer/usr/bin/python3
    Warning: python interpreter not running from IDF_PYTHON_ENV_PATH
    PATH: /Users/USER/.espressif/toUSER/xtensa-esp32-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32-elf/bin:/Users/USER/.espressif/toUSER/xtensa-esp32s2-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32s2-elf/bin:/Users/USER/.espressif/toUSER/xtensa-esp32s3-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32s3-elf/bin:/Users/USER/.espressif/toUSER/riscv32-esp-elf/esp-2021r2-patch3-8.4.0/riscv32-esp-elf/bin:/Users/USER/.espressif/toUSER/esp32ulp-elf/2.28.51-esp-20191205/esp32ulp-elf-binutils/bin:/Users/USER/.espressif/toUSER/esp32s2ulp-elf/2.28.51-esp-20191205/esp32s2ulp-elf-binutils/bin:/Users/USER/.espressif/toUSER/openocd-esp32/v0.11.0-esp32-20220411/openocd-esp32/bin:/Users/USER/esp32/esp-idf/toUSER:/usr/bin:/bin:/usr/sbin:/sbin:/Users/USER/Projects/moddable/build/bin/mac/release
```

</details>

I went to https://github.com/espressif/esp-idf/issues/8760 and applied the recommend fix, to no avail.

```
% pip install "jinja2<3.1"  "itsdangerous<2.1"

Requirement already satisfied: jinja2<3.1 in /Users/USER/.espressif/python_env/idf4.4_py3.10_env/lib/python3.10/site-packages (3.0.3)
Requirement already satisfied: itsdangerous<2.1 in /Users/USER/.espressif/python_env/idf4.4_py3.10_env/lib/python3.10/site-packages (2.0.1)
Requirement already satisfied: MarkupSafe>=2.0 in /Users/USER/.espressif/python_env/idf4.4_py3.10_env/lib/python3.10/site-packages (from jinja2<3.1) (2.1.1)
```

I checked `IDF_PYTHON_ENV_PATH`

```
% echo $IDF_PYTHON_ENV_PATH
/Users/USER/.espressif/python_env/idf4.4_py3.10_env
```

## Questions 

I do not understand why `mcconfig` works and not `node-red-mcu-plugin`. 

Is there something wrong with my configuration?

Thank you for your help!

**Comment by ralphwetzel** [comment]
Thank you for reporting your challenges. Let's try to sort things out step by step:

Initial question: What's the platform you're building on? macOS or Linux or ... ?

First observation: 

> $IDF_PATH identified: /Users/USER/esp32/esp-idf

The statement, that `$IDF_PATH` had to be _identified_, could indicate that it's not defined. Could you please check what `% echo $IDF_PATH` tells you?

**Reply by rei-vilo** [reply]
I am using MacOS 13.0.1 _Ventura_.

`IDF_PATH`gives

```
% echo $IDF_PATH
/Users/USER/.local/share/esp32/esp-idf
```

**Comment by ralphwetzel** [comment]
> Selected board is **esp32/moddable-two**.

Have you been able to run the minimal example successfully on a simulator?

**Reply by phoddie** [reply]
> I ordered the [Moddable Two](https://www.moddable.com/moddable-two.php) board and rushed to test it when I received it. Very nice board with a crisp display.

Thank you! A lot of work went into finding that display.

Good idea to check that `mcconfig` is working correctly. That means all the pieces are present and working. 

@ralphwetzel is the expert on the MCU Plug-in. Since this looks like it might be an IDF configuration issue, I want to ask one question. The IDF needs to be configured in each new terminal window. On macOS, that is usually done by this line in your `.profile` (and something similar on Linux).

```
source $IDF_PATH/export.sh
```

Is it possible that isn't happening in the terminal created by the MCU plugin?

**Reply by rei-vilo** [reply]
Thank you for the answers and hints.

New test: I opened a new Terminal window, ran and then called node-red. So Node-RED is running from the terminal.

Build goes further but fails due to missing "URL" module.

> `### ReferenceError: module "url" not found!`

<details>
  <summary>Log</summary>

```
Starting Build process...
Creating build environment for platform esp32/moddable_two.
Working directory: /Users/USER/.node-red/mcu-plugin-cache/configtgrufvjolkj
$IDF_PATH is defined: /Users/USER/.local/share/esp32/esp-idf
> cd /Users/USER/.node-red/mcu-plugin-cache/configtgrufvjolkj
> source $IDF_PATH/export.sh && mcconfig -d -x localhost:5004 -m -p esp32/moddable_two -t build
Detecting the Python interpreter
Checking "python" ...
Python 3.10.8
"python" has been detected
Adding ESP-IDF toUSER to PATH...
Using Python interpreter in /Users/USER/.espressif/python_env/idf4.4_py3.10_env/bin/python
Checking if Python packages are up to date...
/Users/USER/.local/share/esp32/esp-idf
/Users/USER/.local/share/esp32/esp-idf/toUSER/check_python_dependencies.py
Python requirements from /Users/USER/.local/share/esp32/esp-idf/requirements.txt are satisfied.
Added the following directories to PATH:
/Users/USER/.local/share/esp32/esp-idf/components/esptool_py/esptool
  /Users/USER/.local/share/esp32/esp-idf/components/espcoredump
  /Users/USER/.local/share/esp32/esp-idf/components/partition_table
  /Users/USER/.local/share/esp32/esp-idf/components/app_update
Done! You can now compile ESP-IDF projects.
Go to the project directory and run:
idf.py build
WARNING: Failed to load shell autocompletion for bash version: 3.2.57(1)-release!
# nodered2mcu flows
# xsc main.xsb
# mcrez resources
# xsc flows.xsb
# xsl modules
Total resource size: 0 bytes
# cc mc.resources.c (slots in flash)
ReferenceError: module "url" not found
### ReferenceError: module "url" not found!
make: *** [/Users/USER/Projects/moddable/build/tmp/esp32/moddable_two/debug/configtgrufvjolkj/mc.xs.c] Error 22
```
</details>

So I performed a full refresh of the Moddable SDK and the Node-RED MCU, to no avail.

Where should I look at?

**Reply by rei-vilo** [reply]
> Have you been able to run the minimal example successfully on a simulator?

Selecting simulator with **Simulator / Moddable Two** throws the same error.

> ReferenceError: (host): module "url" not found

<details>
  <summary>Log</summary>

```
Starting Build process...
Creating build environment for platform sim/moddable_two.
Working directory: /Users/OlS/.node-red/mcu-plugin-cache/configtgrufvjolkj
> cd /Users/OlS/.node-red/mcu-plugin-cache/configtgrufvjolkj
> mcconfig -d -x localhost:5004 -m -p sim/moddable_two -t build
# nodered2mcu flows
# xsc main.xsb
# mcrez resources
Total resource size: 0 bytes
# xsc flows.xsb
# xsl modules
# cc mc.resources.c
ReferenceError: (host): module "url" not found
### ReferenceError: (host): module "url" not found!
# exception: (host): module "url" not found!
make: *** [/Users/OlS/Projects/moddable/build/tmp/mac/moddable_two/debug/configtgrufvjolkj/mc.xs.c] Error 22
```

</details>

**Comment by ralphwetzel** [comment]
> Python 3 has been detected but line 99 is hard-coded and calls `python` instead.

This should have been [fixed some time ago](https://github.com/espressif/esp-idf/commit/20487e7a06135db993886129d6316f973f82f7b8) in the Espressif toolchain. I propose to update that as well to the latest version...

**Reply by rei-vilo** [reply]
Indeed, but not in the [recommended](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/devices/esp32.md#installing) 4.4.2 branch.

[Line 99 of export.sh for v4.4.2](https://github.com/espressif/esp-idf/blob/1b16ef6cfc2479a08136782f9dc57effefa86f66/export.sh#L99) still shows

```    python "${IDF_PATH}/tools/check_python_dependencies.py" || return 1```

Should I use another release?

[EDIT] Taking the ESP-IDF master branch seems to work fine.

**Reply by ralphwetzel** [reply]
> Should I use another release?

Looking at the number of changes just around this single topic, I would propose to do it; 4.4.2 definitely has it deficits there.
Yet @phoddie may answer this with a lot more technical background regarding this toolchain...

**Reply by rei-vilo** [reply]
Thank you for the confirmation. 

Using the ESP-IDF master branch, and launching Node-RED from the Terminal so it takes `IDF_PYTHON_ENV_PATH` into account, building still throws the error about `ReferenceError: module "url" not found`.

<details>
  <summary>Log</summary>

```
Starting Build process...
Creating build environment for platform esp32/moddable_two.
Working directory: /Users/USER/.node-red/mcu-plugin-cache/configtgrufvjolkj
$IDF_PATH is defined: /Users/USER/.local/share/esp32/esp-idf
> cd /Users/USER/.node-red/mcu-plugin-cache/configtgrufvjolkj
> source $IDF_PATH/export.sh && mcconfig -d -x localhost:5004 -m -p esp32/moddable_two -t build
Detecting the Python interpreter
Checking "python3" ...
Python 3.10.8
* "python3" has been detected
Adding ESP-IDF toUSER to PATH...
Using Python interpreter in /Users/USER/.espressif/python_env/idf4.4_py3.10_env/bin/python
Checking if Python packages are up to date...
/Users/USER/.local/share/esp32/esp-idf
/Users/USER/.local/share/esp32/esp-idf/toUSER/check_python_dependencies.py
Python requirements from /Users/USER/.local/share/esp32/esp-idf/requirements.txt are satisfied.
Added the following directories to PATH:
/Users/USER/.local/share/esp32/esp-idf/components/esptool_py/esptool
  /Users/USER/.local/share/esp32/esp-idf/components/espcoredump
  /Users/USER/.local/share/esp32/esp-idf/components/partition_table
  /Users/USER/.local/share/esp32/esp-idf/components/app_update
Done! You can now compile ESP-IDF projects.
Go to the project directory and run:
idf.py build
WARNING: Failed to load shell autocompletion for bash version: 3.2.57(1)-release!
# nodered2mcu flows
# xsc main.xsb
# mcrez resources
# xsc flows.xsb
# xsl modules
Total resource size: 0 bytes
# cc mc.resources.c (slots in flash)
ReferenceError: module "url" not found
### ReferenceError: module "url" not found!
make: *** [/Users/USER/Projects/moddable/build/tmp/esp32/moddable_two/debug/configtgrufvjolkj/mc.xs.c] Error 22
make: *** Waiting for unfinished jobs....
```

</details>

To be continued...

**Reply by ralphwetzel** [reply]
> Thank you for the confirmation.

I have to issue a **warning**: My journey to update the Espressif toolchain to 4.4.3 (**today**!) took down my whole development system. Neither `node-red-mcu` nor `Moddable core` built successfully. At one occasion I even met the same error you're facing. I could solve this by doing a fresh installation of Moddable SDK & the Espressif toolchain @ 4.4.2. Your mileage yet may vary ...

**Reply by rei-vilo** [reply]
I am sorry for your bad experience with ESP 4.4.3. Could we consider a container to protect the development system?

Are you referring to the error `ReferenceError: module "url" not found.`?

**Reply by ralphwetzel** [reply]
> I am sorry for your bad experience with ESP 4.4.3.

No problem. I book this into "gathering experience"... 😁 


> Are you referring to the error `ReferenceError: module "url" not found.`?

Exactly. It seems to be related to what (currently) looks like incompatibilities w/ v4.4.3.

**Reply by phoddie** [reply]
Some people at Moddable have been running internally with ESP-IDF 4.4.3 for over two weeks now without a problem. We expect to make that the recommended ESP-IDF release shortly.

The `module "url" not found` is unrelated. @rei-vilo, what is the hash of the Moddable SDK install you are running? Did you install that by following the instructions in the Moddable SDK or, perhaps, by using `xs-dev`?

**Reply by rei-vilo** [reply]
> Exactly. It seems to be related to what (currently) looks like incompatibilities w/ v4.4.3.

I have reinstalled the Moddable and ESP32 SDKS, to no avail.

```
% idf.py --version
ESP-IDF v4.4.2-dirty
```

Unfortunately, the url related error still appears when using Node-RED MCU. `mcconfig` still succeeds.

Again, with so many third-parties tools, assembling a container would be great.

**Reply by rei-vilo** [reply]
> Some people at Moddable have been running internally with ESP-IDF 4.4.3 for over two weeks now without a problem. We expect to make that the recommended ESP-IDF release shortly.
> 
> The `module "url" not found` is unrelated. @rei-vilo, what is the hash of the Moddable SDK install you are running? Did you install that by following the instructions in the Moddable SDK or, perhaps, by using `xs-dev`?

I used this [manual procedure](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/Moddable%20SDK%20-%20Getting%20Started.md#installing). How to get the hash?

**Reply by phoddie** [reply]
Thanks. If you installed manually then you should be at the latest commit, `758b624dc985362c77cd283739bec7f87bc28ac2 `. Here's how to show the hash:

```
cd $MODDABLE
git rev-parse --verify HEAD
```

**Reply by rei-vilo** [reply]
Indeed, 

```
% cd $MODDABLE 
% git rev-parse --verify HEAD
758b624dc985362c77cd283739bec7f87bc28ac2
```

Now, building on Node-RED MCU no longer complains about missing URL but... it seems it does not manage uploading (as it did for the Raspberry Pi Pico). Final lines are 

```
Project build complete. To flash, run this command:
/Users/OlS/.espressif/python_env/idf4.4_py3.10_env/bin/python ../../../../../../../../../.local/share/esp32/esp-idf/components/esptool_py/esptool/esptool.py -p (PORT) -b 460800 --before default_reset --after hard_reset --chip esp32  write_flash --flash_mode dio --flash_size detect --flash_freq 80m 0x1000 build/bootloader/bootloader.bin 0x8000 build/partition_table/partition-table.bin 0x10000 build/xs_esp32.bin
or run 'idf.py -p (PORT) flash'
#
# Built files at /Users/OlS/Projects/moddable/build/bin/esp32/moddable_two/debug/configtgrufvjolkj
#
```

And, either command results in an error `zsh: missing end of string`.

```
% /Users/OlS/.espressif/python_env/idf4.4_py3.10_env/bin/python ../../../../../../../../../.local/share/esp32/esp-idf/components/esptool_py/esptool/esptool.py -p (PORT) -b 460800 --before default_reset --after hard_reset --chip esp32  write_flash --flash_mode dio --flash_size detect --flash_freq 80m 0x1000 build/bootloader/bootloader.bin 0x8000 build/partition_table/partition-table.bin 0x10000 build/xs_esp32.bin
zsh: missing end of string

% idf.py -p (PORT) flash
zsh: missing end of string
```

However, the port `/dev/cu.usbserial-0001` is duly specified on the field.

To be continued...

**Reply by phoddie** [reply]
> Now, building on Node-RED MCU no longer complains about missing URL but

What changed?

**Reply by rei-vilo** [reply]
I did a clean install of everything: Moddable and ESP32 SDKs.

Compared to previous iterations, the difference is the `build` folder of Moddable has been deleted.  

But Node-RED MCU still needs some attention, even with the official Moddable-Two board.

**Reply by phoddie** [reply]
> Compared to previous iterations, the difference is the build folder of Moddable has been deleted.

I see. That's good to know. The [update instructions](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/Moddable%20SDK%20-%20Getting%20Started.md#lin-update) do note that you need to remove the `bin` and `tmp` folders and rebuild tools (step 3 on Linux updates).

> But Node-RED MCU still needs some attention, even with the official Moddable-Two board.

Are you referring to the install issue about about flashing from the MCU Plug-in in the Node-RED Editor?

**Reply by rei-vilo** [reply]
> > Compared to previous iterations, the difference is the build folder of Moddable has been deleted.
> 
> I see. That's good to know. The [update instructions](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/Moddable%20SDK%20-%20Getting%20Started.md#lin-update) do note that you need to remove the `bin` and `tmp` folders and rebuild tools (step 3 on Linux updates).

So those specific instructions are valid for macOS as well :)

> 
> > But Node-RED MCU still needs some attention, even with the official Moddable-Two board.
> 
> Are you referring to the install issue about about flashing from the MCU Plug-in in the Node-RED Editor?

I am referring to the flashing step from Node-RED MCU, which works fine against the Raspberry Pi Pico but not against the ESP32. 

Actually, I was expecting that using the official hardware would bring some ease of use to explore Moddable + Node-RED MCU.

**Reply by ralphwetzel** [reply]
> Some people at Moddable have been running internally with ESP-IDF 4.4.3 for over two weeks now without a problem. We expect to make that the recommended ESP-IDF release shortly.

After several attempts I can confirm meanwhile that v4.4.3 works w/ the Moddable SDK. That's good news! 👍

**Comment by rei-vilo** [comment]
Building succeeds but ends with

```
Project build complete. To flash, run this command:
/Users/USER/.espressif/python_env/idf4.4_py3.10_env/bin/python ../../../../../../../../../.local/share/esp32/esp-idf/components/esptool_py/esptool/esptool.py -p (PORT) -b 460800 --before default_reset --after hard_reset --chip esp32  write_flash --flash_mode dio --flash_size detect --flash_freq 80m 0x1000 build/bootloader/bootloader.bin 0x8000 build/partition_table/partition-table.bin 0x10000 build/xs_esp32.bin
or run 'idf.py -p (PORT) flash'
#
# Built files at /Users/USER/Projects/moddable/build/bin/esp32/moddable_two/debug/configtgrufvjolkj
#
```

Why the port mentioned on the configuration pane isn't taken into account for `(PORT)`? It works fine against the Raspberry Pi Pico.

The suggested command is not valid and should be

```
/Users/OlS/.local/share/esp32/esp-idf/components/esptool_py/esptool/esptool.py --chip esp32 --port /dev/cu.usbserial-0001 --baud 921600 --before=default_reset --after=hard_reset write_flash --flash_mode dio --flash_freq 80m --flash_size 4MB 0x1000 /Users/OlS/Projects/moddable/build/bin/esp32/moddable_two/debug/configtgrufvjolkj/bootloader.bin 0x10000 /Users/OlS/Projects/moddable/build/bin/esp32/moddable_two/debug/configtgrufvjolkj/xs_esp32.bin 0x8000 /Users/OlS/Projects/moddable/build/bin/esp32/moddable_two/debug/configtgrufvjolkj/partition-table.bin
```

The mentioned folder contains

```
% ls /Users/USER/Projects/moddable/build/bin/esp32/moddable_two/debug/configtgrufvjolkj 
bootloader.bin		xs_esp32.a		xs_esp32.map
partition-table.bin	xs_esp32.bin		xs_esp32.sym
```

Still with the same example, sending a timestamp does print anything on the debug node.

<img width="563" alt="204099056-332bf192-b7f7-4306-a1e9-cbcedac7e3e6" src="https://user-images.githubusercontent.com/951776/204384589-126ef784-7cf4-4b0a-98f4-cdab595c6562.png">

To be continued...

**Comment by rei-vilo** [comment]
There is still a major confusing issue on macOS.

If Node-RED is started at boot, build against the ESP32 fails.

> Diagnostic information:
>     IDF_PYTHON_ENV_PATH: (not set)
>     Python interpreter used: /Applications/Xcode.app/Contents/Developer/usr/bin/python3
>     Warning: python interpreter not running from IDF_PYTHON_ENV_PATH

If Node-RED is started from a Terminal window, build succeeds.

> % echo $IDF_PYTHON_ENV_PATH
> /Users/USER/.espressif/python_env/idf4.4_py3.10_env

* Could Node-RED MCU populate `IDF_PYTHON_ENV_PATH` when Node-RED is started at boot?

**Reply by ralphwetzel** [reply]
I'll check...

**Reply by ralphwetzel** [reply]
What did you setup to start Node-RED at boot?

**Reply by rei-vilo** [reply]
The file `/Users/USER/Library/LaunchAgents/org.node-red.agent.plist` launches Node-RED at boot.

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>Label</key>
	<string>org.node-red.agent</string>
	<key>LimitLoadToSessionType</key>
	<string>Aqua</string>
	<key>ProgramArguments</key>
	<array>
	  <string>/usr/local/bin/node</string>
	  <string>/usr/local/lib/node_modules/node-red/red.js</string>
	</array>
	<key>RunAtLoad</key>
	<true/>
	<key>StartInterval</key>
	<integer>3623</integer>
	<key>StandardErrorPath</key>
  <string>/dev/null</string>
	<key>StandardOutPath</key>
  <string>/dev/null</string>
</dict>
</plist>
```

**Comment by rei-vilo** [comment]
# Summary

1. Starting Node-RED at boot ~does not~ works now with release 3.5.0.
2. ~Node-RED needs to be launched from a Terminal window in order to have `IDF_PYTHON_ENV_PATH` defined.~
3. The temporary folder `/Users/USER/Projects/moddable/build/tmp/esp32/moddable_two/debug/` needs to be deleted to get a clean environment.

**UPDATED** 13 Dec 2023 with release 3.5.0

**Reply by phoddie** [reply]
> Starting Node-RED at boot does not work

Interesting. I never do this. `IDF_PYTHON_ENV_PATH` is set by sourcing the ESP-IDF script:

```
source $IDF_PATH/export.sh
```

The set-up documentation ([step 4](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/Moddable%20SDK%20-%20Getting%20Started.md#installing)) recommends putting that line in you `~/.profile` or `~/.zshrc` file so it runs in each new terminal session. Did you do that or is it somehow being ignored / overridden when Node-RED runs at boot?

FWIW - it isn't enough to just set `IDF_PYTHON_ENV_PATH`. Espressif's `export.sh` script does quite a bit, so it really needs to be run.

**Reply by rei-vilo** [reply]
The `~/.zshrc` contains the call to `export.sh`. When opening the Terminal on macOS, it runs and populates the variables ESP-IDF requires.

However, it seems that `~/.zshrc` isn't called when macOS launches Node-RED through `org.node-red.agent.plist`.

So I am no longer using that option. I launch now Node-RED manually from the Terminal to be sure all variables are duly set.

**Reply by phoddie** [reply]
Good to know.

FWIW – the team has implemented a change in the December Moddable SDK release to perform the `source $IDF_PATH/export.sh` automatically when necessary. That might fix the problem with launching Node-RED at book. The December release should be live in the coming days. I think if you Star or Watch the [Moddable SDK](https://github.com/Moddable-OpenSource/moddable) repository, you'll get a notification when it is available.

**Reply by rei-vilo** [reply]
Good to know, I had already stared the Moddable SDK repository!

**Reply by phoddie** [reply]
Version 3.5 of the Moddable SDK contains the changes to automatically `source` as described above. I'm curious if that resolved the problem you were having when launching Node-RED at boot. (The [Getting Started Guide](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/Moddable%20SDK%20-%20Getting%20Started.md) contains instructions on updating, which may be helpful.)

**Reply by rei-vilo** [reply]
New release 3.5 solves the issue. Node-RED can be launched at boot now.

**Reply by phoddie** [reply]
That's great news. Thank you for checking that!

**Reply by rei-vilo** [reply]
Not so fast!

It Terminal answers...

```
% echo $IDF_PYTHON_ENV_PATH
/Users/USER/.espressif/python_env/idf4.4_py3.10_env
```

... Node-RED still complains about

```
Diagnostic information:
Diagnostic information:
    IDF_PYTHON_ENV_PATH: (not set)
    Python interpreter used: /Applications/Xcode.app/Contents/Developer/usr/bin/python3
    Warning: python interpreter not running from IDF_PYTHON_ENV_PATH
    PATH: /Users/USER/.espressif/toUSER/xtensa-esp32-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32-elf/bin:/Users/USER/.espressif/toUSER/xtensa-esp32s2-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32s2-elf/bin:/Users/USER/.espressif/toUSER/xtensa-esp32s3-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32s3-elf/bin:/Users/USER/.espressif/toUSER/riscv32-esp-elf/esp-2021r2-patch3-8.4.0/riscv32-esp-elf/bin:/Users/USER/.espressif/toUSER/esp32ulp-elf/2.28.51-esp-20191205/esp32ulp-elf-binutils/bin:/Users/USER/.espressif/toUSER/esp32s2ulp-elf/2.28.51-esp-20191205/esp32s2ulp-elf-binutils/bin:/Users/USER/.espressif/toUSER/openocd-esp32/v0.11.0-esp32-20220411/openocd-esp32/bin:/Users/USER/esp32/esp-idf/toUSER:/usr/bin:/bin:/usr/sbin:/sbin:/Users/USER/Projects/moddable/build/bin/mac/release
```

**Reply by rei-vilo** [reply]
I added to `~/.node-red/settings.js`

```
process.env.IDF_PYTHON_ENV_PATH = "/Users/USER/.espressif/python_env/idf4.4_py3.10_env"
```

to no avail.

**Reply by phoddie** [reply]
To check that I understand you correctly:

- If you launch Node-RED at start-up, ESP32 builds do work with the plugin
- But... there is a warning about `IDF_PYTHON_ENV_PATH` not being set... and so a warning that an unexpected version of Python is used

**Reply by ralphwetzel** [reply]
`IDF_PYTHON_ENV_PATH` should be defined correctly by `idf_tools.py`, which is always invoked at the beginning of the build process.
I've nevertheless implemented a small change to take over a predefined `IDF_PYTHON_ENV_PATH` value into the build environment; don't know yet if this is a good decision. We'll see...

**Reply by ralphwetzel** [reply]
> I've nevertheless implemented a small change to take over a predefined `IDF_PYTHON_ENV_PATH` value into the build environment; don't know yet if this is a good decision. We'll see...

As a matter of documentation: I've today removed this "fix" (taking over `IDF_PYTHON_ENV_PATH`) again, as it broke the build run w/ ESP-IDF 5.1.1.

**Reply by phoddie** [reply]
Thanks for the note. The ESP-IDF is, unfortunately, a moving target. Updating to IDF v5 was a substantial effort for us, even though the result to most developers using the Moddable SDK looks nearly transparent.
